File: src/validation/areChangeFilesDeleted.ts
import { bulletedList } from '../logging/bulletedList';
import { getChangePath } from '../paths';
import type { BeachballOptions } from '../types/BeachballOptions';
import { getChangesBetweenRefs } from 'workspace-tools';

export function areChangeFilesDeleted(options: Pick<BeachballOptions, 'branch' | 'path' | 'changeDir'>): boolean {
  const { branch, path: cwd } = options;

  const changePath = getChangePath(options);

  console.log(`Checking for deleted change files against "${branch}"`);
  const changeFilesDeletedSinceRef = getChangesBetweenRefs(
    branch,
    'HEAD',
    [
      '--diff-filter=D', // showing only deleted files from the diff.
    ],
    `${changePath}/*.json`,
    cwd
  );

  if (changeFilesDeletedSinceRef.length) {
    const changeFiles = changeFilesDeletedSinceRef.map(file => `- ${file}`);
    console.error(`ERROR: The following change files were deleted:\n${bulletedList(changeFiles)}\n`);
    return true;
  }

  return false;
}

File: rust/src/validation/are_change_files_deleted.rs
use crate::git::commands;
use crate::types::options::BeachballOptions;
use crate::{log_error, log_info};

/// Check if any change files have been deleted (compared to the target branch).
pub fn are_change_files_deleted(options: &BeachballOptions) -> bool {
    let change_path = crate::changefile::read_change_files::get_change_path(options);

    log_info!(
        "Checking for deleted change files against \"{}\"",
        options.branch
    );

    let deleted = commands::get_changes_between_refs(
        &options.branch,
        Some("D"),
        Some("*.json"),
        &change_path,
    )
    .unwrap_or_default();

    if !deleted.is_empty() {
        let items: Vec<&str> = deleted.iter().map(|s| s.as_str()).collect();
        log_error!(
            "The following change files were deleted:\n{}",
            crate::logging::bulleted_list(&items)
        );
    }

    !deleted.is_empty()
}
